### DISCUSSION TITLE
Refactoring enum classes

### DISCUSSION THREAD

**Discussion by ZehMatt** [lief-project/LIEF#826] [General]
In my opinion it would be best to refactor the enum classes such as ELF/enums.hpp PE/enums.hpp etc. to use different names other than their macro counter parts. Things like `MACHINE_TYPES::IMAGE_FILE_MACHINE_UNKNOWN` should be `MachineType::Unknown`. This would prevent the clash with macros and makes it a bit nicer in my opinion, it's a strong enum so we shouldn't use all capital case for such constants and there is definitely no need to prefix them as strong enums dont pollute the global namespace, this would also kill the need to have undef.hpp which can be problematic for when the macro constant is used somewhere else that happens to include something from LIEF. I know this is a bit of API breaking change but it's probably for the better in the long run. Another side effect is that we will finally get rid of the last warning when building LIEF with msvc.

**Comment by romainthomas** [comment]
Hi @ZehMatt,
Yes I agree that the current design of the enums is not the best and the workaround with the undef, not a good solution.
I'm also ok to move to lower case `enum class`. 

If you have some time to address this improvement it would be great!

**Reply by ZehMatt** [reply]
I can try to give it a shot.

**Reply by ZehMatt** [reply]
While starting to work on this I realized theres no .clang-format, considering there will be a massive amount of changes it would be best to get that stone rolling first, a lot of aligned code will no longer be properly aligned and I'm not really feeling too happy about fixing that by hand. I created a .clang-format file that is as close as possible to the current style but there are things where clang-format will simply do things slightly different, there is also no style guide for all cases to be sure since a lot of code has mixed styles. There is also already an open issue for this but it doesn't fully define the goals and how to achieve it project wise.

**Reply by romainthomas** [reply]
Well actually I would like to keep the column alignment as it is since it is (at least for me) more convenient.
I would not be against a `.clang-format` but to be honest I'm the main contributor/maintainer of this project so I would like to keep this style.

**Reply by ZehMatt** [reply]
I understand, clang-format can preserve columns (mostly). Here is the clang-format that is very close the current style: https://gist.github.com/ZehMatt/f23308df3d74a12b64b95b48e07afa01 maybe its close enough and you could just re-format the entire project.
